src/program.rs

```diff
@@ -0,0 +1,315 @@
+                    // edge: last Mid point to Start of remote block
+                    let from = format!(
+                        "\"Mid({block}[{statement}])\"",
+                        block = block_name,
+                        statement = terminator_idx
+                    );
+                    let to = format!("\"Start({next_block}[0])\"", next_block = goto);
+
+                    let from = tables.points.intern(&from);
+                    let to = tables.points.intern(&to);
+
+                    facts.cfg_edge.insert((from, to));
+                }
+            }
+
+            // the most common statement effects: mid point effects
+            for effect in &statement.effects {
+                match effect {
+                    // facts: region_live_at(Region, Point)
+                    Effect::Use { ref region } => {
+                        // region_live_at: a `use` (in Mid points) emits a `region_live_at` fact at Start points
+                        let region = tables.regions.intern(region);
+                        facts.region_live_at.insert((region, start));
+                    }
+
+                    _ => emit_effect(&mut facts, effect, mid, tables),
+                };
+            }
+
+            // commonly used to emit `invalidates` at Start points, like some rustc features do
+            for effect in &statement.effects_start {
+                emit_effect(&mut facts, effect, start, tables);
+            }
+        }
+    }
+
+    Ok(facts.into())
+}
+
+fn emit_effect(facts: &mut Facts, effect: &Effect, point: Point, tables: &mut InternerTables) {
+    match effect {
+        // facts: borrow_region(Region, Loan, Point)
+        Effect::BorrowRegionAt {
+            ref region,
+            ref loan,
+        } => {
+            // borrow_region: a `borrow_region_at` occurs on the Mid point
+            let region = tables.regions.intern(region);
+            let loan = tables.loans.intern(loan);
+
+            facts.borrow_region.insert((region, loan, point));
+        }
+
+        // facts: outlives(Region, Region, Point)
+        Effect::Outlives { ref a, ref b } => {
+            // outlives: a `outlives` occurs on Mid points
+            let region_a = tables.regions.intern(a);
+            let region_b = tables.regions.intern(b);
+
+            facts.outlives.insert((region_a, region_b, point));
+        }
+
+        // facts: killed(Loan, Point)
+        Effect::Kill { ref loan } => {
+            // killed: a loan is killed on Mid points
+            let loan = tables.loans.intern(loan);
+            facts.killed.insert((loan, point));
+        }
+
+        // facts: invalidates(Point, Loan)
+        Effect::Invalidates { ref loan } => {
+            let loan = tables.loans.intern(loan);
+            // invalides: a loan can be invalidated on both Start and Mid points
+            facts.invalidates.insert((point, loan));
+        }
+
+        _ => panic!("Unexpected effect: {:?}", effect),
```

Comment:
I would rather factor `Effect` into nested enums so that this case never happens.